_posts/2019-07-16-stateful-components-in-elm.md

```diff
@@ -0,0 +1,198 @@
+It's often claimed that [Elm] developers should avoid thinking their views as stateful components. While this is indeed a general best design practice, sometimes you may want to make your views reusable (eg. across pages or projects), and if they come with a state... you end up copying and pasting a lot of things.  
+
+We recently published [elm-daterange-picker], a date range picker written in [Elm]. It was the perfect occasion to investigate what a reasonable API for a reusable stateful view component would look like.
+
+![app demo](/assets/2019-07-16-stateful-components-in-elm/demo.gif)
+
+Many component/widget-oriented Elm packages feature a rather raw [Elm Architecture (TEA)] API, directly exposing `Model`, `Msg(..)`, `init`, `update` and `view`, so you can basically import what defines an actual application and embed it within your own application.   
+
+![funny meme](/assets/2019-07-16-stateful-components-in-elm/meme.jpg)
+
+With these, you usually end up writing things like this:
+
+```haskell
+import Counter
+
+
+type alias Model =
+    { widget : Counter.Model
+    , value : Maybe Int
+    }
+
+
+type Msg
+    = CounterMsg Counter.Msg
+
+
+init : () -> ( Model, Cmd Msg )
+init _ =
+    ( { widget = Counter.init, value = Nothing }
+    , Cmd.none
+    )
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case msg of
+        CounterMsg widgetMsg ->
+            let
+                ( newCounterModel, newCounterCommands ) =
+                    Counter.update widgetMsg
+            in
+            ( { model
+                | widget = newCounterModel
+                , value =
+                    case widgetMsg of
+                        Counter.Apply value ->
+                            Just value
+
+                        _ ->
+                            Nothing
+              }
+            , newCommands |> Cmd.map CounterMsg
+            )
+
+
+view : Model -> Html Msg
+view model =
+    div []
+        [ Counter.view model.widget
+            |> Html.map CounterMsg
+        , text (String.fromInt model.value)
+        ]
+```
+
+This certainly works, but let's be frank for a minute and admit this is super verbose and not very developer friendly:
+
+- You need to `Cmd.map` and `Html.map` here and there
+- You need to pattern match `Widget.Msg` to intercept whatever event interests you...
+- ... meaning `Widget` exposes all `Msg`s, which are **implementation details** you now rely on.
+
+There's another way, which [Evan] explained in his now deprecated [elm-sortable-table] package. Among the many good points he has, one idea stroke me as brilliantly simple yet effective to simplify such stateful view components API design:
+
+> **State updates can be managed right from event handlers!**
+
+Let's imagine a simple counter; what if when clicking the *increment* button, instead of calling `onClick` with some `Increment` message, we would call **a user-provided one** with the new counter state updated accordingly?
```

Comment:
**a user-provided one** what? You mean **a user-provided function**?

Reply:
A user-provided message. To me it's rather clear, but for other readers I don't know. How would you rephrase it?